File: src/main/java/withoutaname/mods/immersivesignals/modules/signal/items/SignalItem.java
package withoutaname.mods.immersivesignals.modules.signal.items;

import javax.annotation.Nonnull;

import net.minecraft.core.BlockPos;
import net.minecraft.util.Mth;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.level.Level;
import withoutaname.mods.immersivesignals.modules.signal.blocks.BaseSignalBlock;
import withoutaname.mods.immersivesignals.setup.ModSetup;

public class SignalItem extends Item {
	
	public static final String REGISTRY_NAME = "signal";
	
	public SignalItem() {
		super(new Item.Properties()
				.stacksTo(64)
				.tab(ModSetup.DEFAULT_CREATIVE_TAB));
	}
	
	@Nonnull
	@Override
	public InteractionResult useOn(UseOnContext context) {
		Level level = context.getLevel();
		if (!level.isClientSide) {
			BlockPos pos = level.getBlockState(context.getClickedPos()).getMaterial().isReplaceable() ? context.getClickedPos() : context.getClickedPos().relative(context.getClickedFace());
			int rotation = Mth.floor((context.getRotation() * 16.0F / 360.0F) + 0.5F) & 15;
			if (BaseSignalBlock.createSignal(level, pos, rotation, 5, true, true)) {
				context.getItemInHand().shrink(1);
				return InteractionResult.SUCCESS;
			}
		}
		return super.useOn(context);
	}
	
}

File: src/main/kotlin/eu/withoutaname/mod/immersivesignals/signal/item/SignalItem.kt
package eu.withoutaname.mod.immersivesignals.signal.item

import eu.withoutaname.mod.immersivesignals.setup.ModSetup
import eu.withoutaname.mod.immersivesignals.signal.block.SignalBlock
import eu.withoutaname.mod.immersivesignals.signal.util.SignalConfiguration
import net.minecraft.item.Item
import net.minecraft.item.ItemUseContext
import net.minecraft.util.ActionResultType
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import kotlin.math.round

object SignalItem : Item(ModSetup.DEFAULT_ITEM_PROPERTIES.stacksTo(1)) {
    override fun useOn(pContext: ItemUseContext): ActionResultType {
        val level: World = pContext.level
        if (!level.isClientSide) {
            val pos: BlockPos = if (level.getBlockState(pContext.clickedPos).material
                .isReplaceable
            ) pContext.clickedPos else pContext.clickedPos.relative(pContext.clickedFace)
            val rotation: Int = (round((180 - pContext.rotation) * 16.0f / 360.0f).toInt() + 16) % 16
            if (SignalBlock.createSignal(level, pos, SignalConfiguration(rotation, 5.5f))) {
                pContext.itemInHand.shrink(1)
                return ActionResultType.SUCCESS
            }
        }
        return super.useOn(pContext)
    }
}
